### PR TITLE
fix am32 flashing error when not all 4 ESCs are connected

### PR DESCRIPTION
**Pull request by kikoqiu**
Fix for the following errors.
When only connect one ESC to fc, the flashing process will fail.
When flashing am32 for the first time, it will fail even with ignore mcu & layout.

### CODE REVIEW

File: src/utils/FourWay.js [hunk 1]
```diff
@@ -1153,11 +1153,17 @@ class FourWay {
...
+        //call initFlash before call to read, or the read will read from the last (the 4th) esc
+        await this.initFlash(target);
         const newVectorStartBytes = flash.subarray(firmwareStart, firmwareStart + 4);
         const currentVectorStartBytes = (await this.read(firmwareStart, 4, 10)).params;
 
         if (!compare(newVectorStartBytes, currentVectorStartBytes)) {
-          throw new InvalidHexFileError('Invalid hex file');
+          if(!force){//only thrown error with force
+            throw new InvalidHexFileError('Invalid hex file');
+          }else{
```
Comment (stylesuxx) [inline src/utils/FourWay.js:1164]: Please remove else branch.

File: src/utils/FourWay.js [hunk 2]
```diff
@@ -1153,11 +1153,17 @@ class FourWay {
...
          * different for different MCUs
          */
+        //call initFlash before call to read, or the read will read from the last (the 4th) esc
+        await this.initFlash(target);
         const newVectorStartBytes = flash.subarray(firmwareStart, firmwareStart + 4);
         const currentVectorStartBytes = (await this.read(firmwareStart, 4, 10)).params;
 
         if (!compare(newVectorStartBytes, currentVectorStartBytes)) {
-          throw new InvalidHexFileError('Invalid hex file');
+          if(!force){//only thrown error with force
```
Comment (stylesuxx) [inline src/utils/FourWay.js:1162]: add space after if and opening bracket. Remove comment - implicit by if clause.

File: src/Containers/App/index.jsx [hunk 3]
```diff
@@ -283,7 +283,7 @@ class App extends Component {
       if(result) {
         result.index = target;
         result.ref = React.createRef();
-        individual[target] = result;
+        individual[i] = result;
```
Comment (stylesuxx) [inline src/Containers/App/index.jsx:286]: Careful: `i` is the position in the target array, not in the individual array. Here we should find the index of the individual ESC matching the actual target.

**Review by stylesuxx — CHANGES_REQUESTED** [review]

File: src/Containers/App/index.jsx (hunk 3 above)
Comment (kikoqiu) [reply src/Containers/App/index.jsx:286]: I see. Here I should use `escs.individual.findIndex((esc) => esc.index === target)`. I will update soon after I hava this tested.

File: src/utils/FourWay.js [hunk 4]
```diff
@@ -1153,11 +1153,15 @@ class FourWay {
...
          * different for different MCUs
          */
+        //call initFlash before call to read, or the read will read from the last (the 4th) esc
+        await this.initFlash(target);
         const newVectorStartBytes = flash.subarray(firmwareStart, firmwareStart + 4);
         const currentVectorStartBytes = (await this.read(firmwareStart, 4, 10)).params;
 
         if (!compare(newVectorStartBytes, currentVectorStartBytes)) {
-          throw new InvalidHexFileError('Invalid hex file');
+          if (!force) {
```
Comment (stylesuxx) [inline src/utils/FourWay.js:1162]: Please move the force check into the if above, should be the first check to be evaluated.

**Review by stylesuxx — CHANGES_REQUESTED** [review]

File: src/utils/FourWay.js (hunk 4 above)
Comment (kikoqiu) [reply src/utils/FourWay.js]: Done. I disabled the check when it's in force mode.

**Review by stylesuxx — APPROVED** [review]

**Comment by stylesuxx** [comment]
Awesome, thank you so much. Will test it later on and merge. Thanks again for your contribution!

**Review by stylesuxx — CHANGES_REQUESTED** [review]

File: src/Containers/App/index.jsx [hunk 5]
```diff
@@ -283,7 +283,8 @@ class App extends Component {
       if(result) {
         result.index = target;
         result.ref = React.createRef();
-        individual[target] = result;
+        const targetEscInIndividual = escs.individual.find((esc) => esc.index === target);
+        individual[targetEscInIndividual] = result;
```
[the remark this answers is not in the document]
Comment (kikoqiu) [reply src/Containers/App/index.jsx:287]: TargetEscInIndividual.index equals to target.index as the original code.
I change it becase it brings an issue when not all 4 escs are present. 

For example, When 
individual=[{index:0},{index:3}]
target=[{index:3}]

if we use `target.index` here, individual will be `[{index:0},{index:3},undefined,{index:3}]` at last.

Here we use `targetEscInIndividual=individual.find((esc) => esc.index === target)`, which equals to 1.
And we can get `individual[1]={index:3}`.

[the remark this answers is not in the document]
Comment (stylesuxx) [reply src/Containers/App/index.jsx:287]: Just to recap and sanity check:

`individual` contains only the actually available ESCs. When initially reading the ESC ist only pushed to this array if it has been read successful. The ESC itself contains the actual Index, so `esc.index` does not equal the array index in `individual`.

`target` holds all the ESCs to be flashed - realistically speaking this is either one or all.

So to know which ESC in the individual array needs to be updated we need to compare the target ESCs index against the items in the individual array to get the correct position in the individual array.

So it should actually look like this:

```
const individualIndex = escs.individual.findIndex((esc) => esc.index === target);
individual[individualIndex] = result;
```

Sorry for the confusion.

**Review by stylesuxx — APPROVED** [review]

**Pull request merged**